### PR TITLE
Add post_dataloading_processing method to Trainer

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #2002
* #2001
* #1995
* __->__ #1985

We are adding more actions to convert the raw inputs and label.

1. The new CP can do the input/label/BlockMask sharding this in this method.
2. The experimental full dtensor model can simply override this method without changing too many Trainer code.

This method is extracted from https://github.com/pytorch/torchtitan/pull/1857

Makeing this a standalone PR allows us to continue the two projects above without one blocks another.

### CODE REVIEW

File: torchtitan/train.py [hunk 1]
```diff
@@ -410,26 +410,37 @@ def batch_generator(
...
 
         if getattr(self.model_args, "use_flex_attn", False):
-            extra_kwargs["attention_masks"] = model_parts[0].get_attention_masks(
+            extra_kwargs["attention_masks"] = self.model_parts[0].get_attention_masks(
                 input_batch=inputs,
                 tokenizer=self.tokenizer,
                 extra_inputs=extra_inputs,
             )
 
+        return inputs, label, extra_inputs, extra_kwargs
```
Comment (tianyu-l) [inline torchtitan/train.py]: I think we should add docstring for returns, especially on the difference between `extra_inputs` and `extra_kwargs`.

Also not sure if we should just merge `inputs` and `extra_inputs`. Not urgent though.

File: torchtitan/train.py [hunk 2]
```diff
@@ -410,26 +410,37 @@ def batch_generator(
 
             yield input_dict, labels
 
-    def forward_backward_step(
-        self, input_dict: dict[str, torch.Tensor], labels: torch.Tensor
-    ) -> torch.Tensor:
-        model_parts = self.model_parts
-        parallel_dims = self.parallel_dims
-
+    def post_dataloading_processing(
```
Comment (tianyu-l) [inline torchtitan/train.py]: This name is accurate in where it should be called, but we are putting it not right after dataloading. Rather we are putting it before training, which makes sense because when other library depends on torchtitan training but not torchtitan data loading, this is the right place to put it.

I just wonder if we could have another name that can express it's happening right before (but mostly as part of) the training, e.g. a bad and verbose version would be `pre-actual-training-last-minute-data-preparation`.

Comment (fegin) [reply torchtitan/train.py]: How about `pre_training_data_processing` or `pre_training_data_preparation` or if the "last" is really an important message, then `final_data_preparation`?

Comment (tianyu-l) [reply torchtitan/train.py]: I was trying to avoid the term "pre-training" which could cause confusion.

I think we can go with `post_dataloading_process`, seems no ambiguity.
```suggestion
    def post_dataloading_process(
```

**Review by tianyu-l — APPROVED** [review]
